### ISSUE TITLE
How are results cleaned up?

### ISSUE THREAD

**Issue by brianwawok** [celery/celery#4660]
How and when do results get cleaned up?  I see the **result_expires** setting and talk about a daily 4am celery beat task.. is that the only way we delete results?  There is no concept of "we already used these results, we can discard them now"

Specifically asking about the redis backend. As I use celery for more and more tasks, it slowly uses more and more memory, and I keep raising the memory limit of my redis. I either need to make results get cleaned up "smarter", or I need to start using the database to store results in. Redis seems not a great tool to store an unbounded collection of stuff that is only cleaned up once per day....

**Comment by georgepsarakis** [comment]
@brianwawok you can use the the `forget` method of the result. For details see [here](https://celery.readthedocs.io/en/latest/reference/celery.result.html#celery.result.AsyncResult.forget).

**Comment by brianwawok** [comment]
@gcheshkov  ok so let's say this is a chord..

I go and make 20 tasks with ignore_result=False

So in the final step of my Chord, I go spin through all of my task_results and call forget() on each one?

That is interesting....

**Comment by brianwawok** [comment]
@georgepsarakis I don't think this works for chords... shouldn't a chord automatically "forget" all of the results he used?

**Comment by georgepsarakis** [comment]
Can you please construct an example, so we can examine it more closely?

**Comment by brianwawok** [comment]
@georgepsarakis yes here is the basic idea...

```
@app.task(ignore_result=False)
def do_some_bulk_processing(chunk):
    errors = []
    updates = 0
    # do some stuff
    return errors, updates

@app.task()
def collect_results(task_results):
    updated = 0
    errors = []
    if not isinstance(task_results[0], tuple) and not isinstance(task_results[0], list):
        task_results = [task_results]
    for task_updated, task_errors in task_results:
        updated += task_updated
        errors.extend(task_errors)
    # send out an email, you updated 22333344 items, and had these errors: [x, y, z]
    # TODO at this point I do not care about the results for the 10,000 tasks of do_some_bulk_processing, but I see no way to "forget" them
    
def process_bulk_upload(some_upload_file):
    raw_data = _parse_file_into_array(some_upload_file)
    tasks = [do_some_bulk_processing.s(r) for r in raw_data]
    chord(tasks)(collect_results.s())
```

Two things seem like they could help this.

1) A way to trigger a "forget" when I hit the TODO piece.. assuming redis has forget coded into it.. I haven't looked super deep, but it looks like you guys generally use auto expire, which isn't what we want here.

2) Really, once a chord is done with his subtasks.. why not automatically "forget" that work?  No reason to just hang onto it for a day or whatever?

**Comment by georgepsarakis** [comment]
Redis backend seems to be [expiring chord header tasks](https://github.com/celery/celery/blob/master/celery/backends/redis.py#L278) when configured, see also [here](https://github.com/celery/celery/blob/master/celery/backends/redis.py#L110). 

I am not really sure, but if we call `.forget` [here](https://github.com/celery/celery/blob/master/celery/backends/base.py#L436), it will achieve what you are describing:

```
def apply_chord(self, header_result, body, **kwargs):
        self.ensure_chords_allowed()
        self.fallback_chord_unlock(header_result, body, **kwargs)
        header_result.forget()
```

Maybe there is consequences that we haven't thought of though, for example what happens with error handling & retries. If you want to test this hypothesis, please let me know about the results :smile: !

**Comment by jenstroeger** [comment]
Helpful discussion, thanks, particulalry in the context of https://github.com/celery/celery/issues/4642. While we’re still tied to Celery I need to remove results from the Redis store manually and from @georgepsarakis’s comment it sounds like [`forget()`](https://celery.readthedocs.io/en/latest/reference/celery.result.html#celery.result.AsyncResult.forget) should work.
